### PR TITLE
libkb.HTTPSrv: fix a potential race; idempotent EnsureActive(); reuse ServeMux

### PR DESCRIPTION
**Pull request by songgao**
If `Addr()` is called immediately after `Start()` is called, it's possible that `Addr()` grabs lock before the goroutine spawned in `Start()` does, and it'd find `active` to be false, and returns an error. So this PR sets `active` to true before releasing the lock in `Start()`. (I don't have proof from the logs to show this is indeed what happened for Patrick and Miles, but this is the likely cause that I suspect.)

Also changing `Start()` into an idempotent `EnsureActive()`.

### CODE REVIEW

File: go/libkb/httpsrv.go [hunk 1]
```diff
@@ -125,6 +124,19 @@ func (h *HTTPSrv) Start() (err error) {
 	return nil
 }
 
+// EnsureActive is like Start, except that in case the server is already active
+// at the time this is called, it returns a nil error rather than
+// errHTTPServerAlreadyRunning.
+func (h *HTTPSrv) EnsureActive() (err error) {
```
Comment (jzila) [inline go/libkb/httpsrv.go]: You could override `Start` to do the right thing.

File: go/libkb/httpsrv.go [hunk 2]
```diff
@@ -110,10 +110,9 @@ func (h *HTTPSrv) Start() (err error) {
 		Addr:    address,
 		Handler: h.ServeMux,
 	}
+	// Set active to true before returning.
+	h.active = true
```
Comment (jzila) [inline go/libkb/httpsrv.go]: Where is this used?

Comment (songgao) [reply go/libkb/httpsrv.go]: In `Active()`, it uses this field to determine if the server is up.

File: go/libkb/httpsrv.go (hunk 1 above)
Comment (songgao) [reply go/libkb/httpsrv.go]: It's also used in chat, and I was not sure if Mike is OK with changing. But I guess I'll push a follow-up commit and see what Mike thinks. We can also just move this type into libhttpserver.

**Comment by songgao** [comment]
Removing reviewers for now; I'm gonna make a couple more changes ...

**Comment by songgao** [comment]
Ready for review again!

**Comment by mmaxim** [comment]
I'm not very excited about how big this change is, what exactly is the problem KBFS is having with it? Why don't I see anything like it in chat?

**Comment by songgao** [comment]
I think the reason why chat doesn't have any issues with this is because in chat `Addr()` is only called when user tries to see the attachment content, and that once FE has the address, it'd keep using that until next time the process restarts. In KBFS, we call GetHTTPAddressAndToekn (which calls `Addr()`) in two places: 1) as soon as app starts; 2) when an error happens in webView indicating that token is invalid, which means 1) we call `Addr()` earlier on than chat, and 2) we may call it at some unfortunate times that chat isn't.

That being said, I don't have direct proof that shows this is indeed the cause for the bug hit by Patrick and Miles. I do know the error is "server not running".

**Comment by mmaxim** [comment]
`Addr` doesn't even reference `active`?

**Comment by mmaxim** [comment]
Seems like just moving `active = true` out of that goroutine should be good enough?

**Comment by songgao** [comment]
> Addr doesn't even reference active?

Ah you're right. `Addr()` is actually race-free. It must be something else then. Sorry for the wrong fix!

**Comment by mmaxim** [comment]
It might be you are racing Start and Addr on foreground

**Comment by songgao** [comment]
Yeah it might be Addr() was called before appstate change gets delivered, so it grabbed lock before Start() could.

**Pull request closed without merging**